X-Withdrawn/RFC0043-Loading-Module-Into-Isolated-AssemblyLoadContext.md

```diff
@@ -0,0 +1,183 @@
+Otherwise, the type resolution will search the default-ALC only.
+
+So in the above imagined scenario, it's possible to make it work as expected if you construct the object using `[M\B::C]` from the default-ALC.
+But it's non-intuitive and I think many existing scripts may fail when modules started to be loaded in separate load contexts.
+
+### Caching Problem
+
+PowerShell has a lot caches related to `System.Type`.
+Some have `Type` as the value, such as the cache in `TypeResolver` and the type cache for `TypeNameAst`.
+Some have `Type` as the key, such as the member tables in `DotNetAdapter` and the binder caches in DLR.
+
+For the former category, the cache data structure needs to be updated so as to make sure the cache is associated with a load context.
+So for example, when type resolution happens for `[C]` that is running from an isolated module M,
+then `TypeResolver` should use the cache associated with ALC-M.
+
+Another example, if a `TypeNameAst` has resolve the type via reflection in default-ALC,
+and the `ScriptBlock` containing that `TypeNameAst` is being executed in the context of an isolated module M,
+then the cached type referred by `TypeNameAst` should be voided and another resolution should be triggered.
+Similarly, if the `ScriptBlock` has already been compiled into LINQ expression tree and have the delegate generated,
+then the cached delegate needs to be voided and another compilation needs to be enforced to make sure type resolution is correct.
+
+For the latter category, theoretically they will continue to work without change,
+because types loaded into different load contexts are different.
+Initially, I think we still need to replace the cache data structure with something like `ConditionalWeakTable<AssemblyLoadContext, ...>`,
+so that a load context can be GC collected after the corresponding module is removed.
+However, it turns out it's unlikely that the module load contexts can be reclaimed at all in PowerShell.
+We will visit that topic in the next section.
+
+As you can imagine, it will be a huge work item to just getting the current caching correct.
+And besides, the bookkeeping to make sure caches from the first category are properly voided and rebuilt will be expensive and hurt performance.
+
+### Cannot GC Module's AssemblyLoadContext
+
+.NET Core 3.0 supports unloading an `AssemblyLoadContext` and all assemblies in it.
+Ideally, we would like to reclaim the load context associated with an isolated module when the module is removed.
+However, it's very likely not possible no matter how we update our caches to avoid holding an `AssemblyLoadContext` instance indefinitely.
+This is because Dynamic Language Runtime (DLR) also holds types in its cache, indefinitely.
+
+The DLR types live in the default-ALC, and it will generate a `DynamicMetaObject` object for a late-binding operation,
```

Comment:
For runspaces we could have local types and local caches - cache searching could be thread distributed. If such cache is a ConditionalWeakTable perhaps it will is still possible to unload custom load context.